### ISSUE TITLE
unexpected ChannelClosedExceptions

### ISSUE THREAD

**Issue by wmorgan** [BuoyantIO/linkerd#482]
(Feel free to suggest a better title. Just trying to capture the details on slack.)
(May or may not be related to #405.)

@jacob-koren reports:

```Client:
org.databene.contiperf.PerfTestExecutionError: java.lang.AssertionError: Error message from HttpResponseException not matches: 502--ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4

    at org.databene.contiperf.util.ContiPerfUtil.executionError(ContiPerfUtil.java:58)
    at org.databene.contiperf.junit.JUnitInvoker.invoke(JUnitInvoker.java:46)
    at org.databene.contiperf.util.InvokerProxy.invoke(InvokerProxy.java:38)
    at org.databene.contiperf.PerformanceTracker.invoke(PerformanceTracker.java:87)
    at org.databene.contiperf.TimedRunner.run(TimedRunner.java:44)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: Error message from HttpResponseException not matches: 502--ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4
    at org.junit.Assert.fail(Assert.java:88)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at com.zooz.smartrouting.utils.OperationsUtil.undeployAndGetToEnvAndAssert(OperationsUtil.java:77)
    at com.zooz.smartrouting.api.ConcurrentConfigurationOperationsFlow.runTest(ConcurrentConfigurationOperationsFlow.java:206)
    at com.zooz.smartrouting.api.ConcurrentConfigurationOperationsFlow.testDifferentFlowOwnerDeployingAndUndeployingInParallal(ConcurrentConfigurationOperationsFlow.java:172)
    at sun.reflect.GeneratedMethodAccessor18.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.databene.contiperf.junit.JUnitInvoker.invoke(JUnitInvoker.java:43)
    ... 4 more

Server:
E 0615 15:35:20.500 THREAD25: service failure
com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /10.0.3.170:22393 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /10.0.5.108:21738, Upstream Client Id: Not Available, Downstream Address: /10.0.3.170:22393, Downstream Client Id: #/io.l5d.marathon/sre-testing, Trace Id: 6bc78e8f619f66f4.8633598551f506cf<:6bc78e8f619f66f4
    at com.twitter.finagle.NoStacktrace(Unknown Source)
```

**Comment by wmorgan** [comment]
In the absence of being able to repro on our end, we would like a tcpdump from @jacob-koren to diagnose.

**Comment by jacob-koren** [comment]
I uploaded the tcpdump in slack channel

**Comment by wmorgan** [comment]
@jacob-koren Thanks, we're taking a look at this now. It looks like the pcap file doesn't include linkerd -> server traffic, which we'd like to take a look at. Are you able to capture that traffic as well? (If it's helpful, we can give you a pcap command if you tell us the ports of the upstream service -- since they're on Mesos it will be something weird. Maybe 22393 in the example above?)

**Comment by wmorgan** [comment]
The other thing that might be useful is some stats dumps so we can get a sense of # nodes, how long connections live, etc. You can run this small script when generating the errors:

```
#!/bin/bash

if [ !-e curl ]; then;
  # FIXME: write a version that can work with either wget or curl or nc
  echo "curl does not exist, can't fetch stats. ask @stevej to give you a version that uses wget"
  exit -1
done

while true; do
  curl -s http://localhost:9990/admin/metrics.json > l5d_metrics_`date -u +'%s'`.json
  sleep 60;
done
```

Sorry for all the requests. It's not easy to debug this sort of thing remotely!

**Comment by jacob-koren** [comment]
https://linkerd.slack.com/files/jacobk/F1HBGUH6F/linkerd.pcap

**Comment by jacob-koren** [comment]
[Archive.zip](https://github.com/BuoyantIO/linkerd/files/317802/Archive.zip)

**Comment by jacob-koren** [comment]
Added stats of about 10 minutes

**Comment by olix0r** [comment]
Thanks for the extra data, @jacob-koren. This helps us understand the failure in more detail.  From what I gather:
- The sre-testing service has only a single node
- There are up to ~3 connections to this node at a given point in time. (By default, linkerd provisions connections as needed to be able to satisfy all requests).
- The service occasionally emits a 400 response (_No flow is deployed for ..._).
  - When it returns a 400, the server signals `Connection: close` on the response and immediately closes the connection: ![screen shot 2016-06-16 at 3 32 29 pm](https://cloud.githubusercontent.com/assets/240738/16135345/e186755c-33d7-11e6-8bc7-8074bee906e7.png)
 - The next on this service sometimes (not always) fails with a connection error.

This may be a timing issue in linkerd where the connection is not marked as closed before new requests are dispatched onto the connection.  I think I have enough data to build a good reproduction we can test against.  Thanks!

**Comment by olix0r** [comment]
I've been able to reproduce this locally:

- Run a [small web server that closes connections](https://gist.github.com/olix0r/872f206ed22fce058fcd67c7fdc52e65)
  - `go run shutter.go`
- run `./sbt linkerd-examples/http:run`
  - `echo "127.1 8080" >linkerd/examples/io.l5d.fs/ugly`
- in two terminals run: `while true ; do curl -H 'Host: ugly' localhost:4140 ; done`

After a few seconds, the following error is encountered in linkerd's logs:
```
om.twitter.finagle.ChannelClosedException: null at remote address: /127.0.0.1:8080 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /127.0.0.1:62989, Upstream Client Id: Not Available, Downstream Address: /127.0.0.1:8080, Downstream Client Id: #/io.l5d.fs/ugly, Trace Id: 525ae9bf20f23725.e45f5414a27b8f9b<:525ae9bf20f23725
	at com.twitter.finagle.NoStacktrace(Unknown Source)
Caused by: java.nio.channels.ClosedChannelException
```

This should give us enough information to debug this in linkerd.  Thanks again for all your help documenting the problem, @jacob-koren.

**Comment by olix0r** [comment]
As described in https://github.com/twitter/finagle/issues/517, I've isolated this to a problem--probably a race condition--in finagle's Netty3 integration.  I have a branch that enables Netty4 which seems to fix this issue:

Binary here: https://slack-files.com/T0JV2DX9R-F1J07MA0M-267f77a77f

To enable netty4, linkerd needs a router config like:
```yaml
- protocol: http
  ...
  servers:
  - port: 4140
    ip: 0.0.0.0
    engine:
      kind: netty4
  client:
    engine:
      kind: netty4
```

This feature is expected to be released in linkerd-0.7.0.  Let me know if it resolves the issue for you--if it does, I may just make netty4 the default engine in 0.7.0...

**Comment by olix0r** [comment]
Sorry, please disregard that. I've reproduced this problem with netty4 as well.  I'm working with the finagle team to fix this...

**Comment by olix0r** [comment]
Alright, I have a build against a patched version of finagle that has fixes for a related issue. The binary is available at https://slack-files.com/T0JV2DX9R-F1JL2EN78-3a17e49419

We're still working through some of the details of the fix, but I expect that the provided binary does not exhibit the exceptional behavior.  When you get a moment, @jacob-koren, I'd be curious to know if this resolves your issues. Thanks!

**Comment by jacob-koren** [comment]
Do you have the source? I need to build a docker image to use  in dc/os

**Comment by olix0r** [comment]
sorry @jacob-koren -- yeah the build is against a development version of finagle using these two branches: https://github.com/BuoyantIO/linkerd/tree/ver%2Fnetty4 https://github.com/twitter/finagle/pull/518 -- it's somewhat tricky to get all of the unreleased dependencies in order, so I've pushed an image to [buoyantio/linkerd:gh482](https://hub.docker.com/r/buoyantio/linkerd/tags/).  Let me know if this doesn't work as expected for any reason.

**Comment by jacob-koren** [comment]
Yes. It's fixed!

**Comment by olix0r** [comment]
excellent news! I expect that this fix will make it into 0.7.0.  We'll publish a release once all of these fixes have made it into a stable finagle release.
